File: backend/src/modules/batch/batch.service.ts
import prisma from '../../infra/prisma.js';
import { onNoticeCreated } from '../notification/notification.hooks.js';

// ─── DTOs ────────────────────────────────────────────────────────────

export interface CreateBatchDto {
    name: string;
    subject?: string;
    description?: string;
    startTime?: string;
    endTime?: string;
    days?: string[];
    maxStudents?: number;
}

export interface UpdateBatchDto {
    name?: string;
    subject?: string;
    description?: string;
    startTime?: string;
    endTime?: string;
    days?: string[];
    maxStudents?: number;
    status?: string;
}

// Reusable include for batch queries — keeps all queries consistent
const BATCH_LIST_SELECT = {
    id: true,
    name: true,
    subject: true,
    description: true,
    startTime: true,
    endTime: true,
    days: true,
    maxStudents: true,
    status: true,
    createdAt: true,
    updatedAt: true,
    _count: { select: { members: true, notes: true, notices: true } },
    members: {
        where: { role: 'TEACHER' },
        take: 1,
        select: {
            id: true,
            role: true,
            member: {
                select: {
                    id: true,
                    user: { select: { id: true, name: true, picture: true } },
                },
            },
        },
    },
} as const;

const BATCH_DETAIL_SELECT = {
    id: true,
    coachingId: true,
    name: true,
    subject: true,
    description: true,
    startTime: true,
    endTime: true,
    days: true,
    maxStudents: true,
    status: true,
    createdAt: true,
    updatedAt: true,
    _count: { select: { members: true, notes: true, notices: true } },
    members: {
        select: {
            id: true,
            role: true,
            createdAt: true,
            member: {
                select: {
                    id: true,
                    role: true,
                    user: { select: { id: true, name: true, picture: true, email: true } },
                    ward: { select: { id: true, name: true, picture: true } },
                },
            },
        },
        orderBy: [{ role: 'asc' as const }, { createdAt: 'asc' as const }],
    },
};

// ─── Service ─────────────────────────────────────────────────────────

export class BatchService {

    // ── CRUD ──────────────────────────────────────────────────────────

    async create(coachingId: string, data: CreateBatchDto) {
        return prisma.batch.create({
            data: { coachingId, ...data },
            select: BATCH_LIST_SELECT,
        });
    }

    async list(coachingId: string, status?: string) {
        return prisma.batch.findMany({
            where: {
                coachingId,
                ...(status ? { status } : {}),
            },
            select: BATCH_LIST_SELECT,
            orderBy: { createdAt: 'desc' },
        });
    }

    async getById(batchId: string) {
        return prisma.batch.findUnique({
            where: { id: batchId },
            select: BATCH_DETAIL_SELECT,
        });
    }

    async update(batchId: string, data: UpdateBatchDto) {
        return prisma.batch.update({
            where: { id: batchId },
            data,
            select: BATCH_LIST_SELECT,
        });
    }

    async delete(batchId: string) {
        return prisma.batch.delete({ where: { id: batchId } });
    }

    // ── Members ───────────────────────────────────────────────────────

    async addMembers(batchId: string, memberIds: string[], role: string = 'STUDENT') {
        // Use createMany with skipDuplicates for idempotent adds
        const result = await prisma.batchMember.createMany({
            data: memberIds.map(memberId => ({
                batchId,
                memberId,
                role,
            })),
            skipDuplicates: true,
        });
        return result;
    }

    async removeMember(batchMemberId: string) {
        return prisma.batchMember.delete({ where: { id: batchMemberId } });
    }

    async getMembers(batchId: string) {
        return prisma.batchMember.findMany({
            where: { batchId },
            select: {
                id: true,
                role: true,
                createdAt: true,
                member: {
                    select: {
                        id: true,
                        role: true,
                        user: { select: { id: true, name: true, picture: true, email: true } },
                        ward: { select: { id: true, name: true, picture: true } },
                    },
                },
            },
            orderBy: [{ role: 'asc' }, { createdAt: 'asc' }],
        });
    }

    /** Get coaching members NOT already in this batch — for the "Add Members" picker */
    async getAvailableMembers(coachingId: string, batchId: string, role?: string) {
        const existingMemberIds = await prisma.batchMember.findMany({
            where: { batchId },
            select: { memberId: true },
        });
        const ids = existingMemberIds.map(m => m.memberId);

        return prisma.coachingMember.findMany({
            where: {
                coachingId,
                status: 'active',
                id: { notIn: ids },
                ...(role ? { role } : {}),
            },
            select: {
                id: true,
                role: true,
                user: { select: { id: true, name: true, picture: true } },
                ward: { select: { id: true, name: true, picture: true } },
            },
            orderBy: { createdAt: 'asc' },
        });
    }

    // ── Notes (Study Material) ────────────────────────────────────────

    async createNote(batchId: string, uploadedById: string, data: {
        title: string;
        description?: string;
        attachments?: { url: string; fileName?: string; description?: string; fileType?: string; fileSize?: number; mimeType?: string }[];
    }) {
        const { attachments, ...noteData } = data;
        return prisma.batchNote.create({
            data: {
                batchId,
                uploadedById,
                ...noteData,
                ...(attachments && attachments.length > 0 ? {
                    attachments: {
                        create: attachments.map(a => ({
                            url: a.url,
                            fileName: a.fileName ?? null,
                            description: a.description ?? null,
                            fileType: a.fileType || 'pdf',
                            fileSize: a.fileSize || 0,
                            mimeType: a.mimeType ?? null,
                        })),
                    },
                } : {}),
            },
            include: {
                uploadedBy: { select: { id: true, name: true, picture: true } },
                attachments: { orderBy: { createdAt: 'asc' } },
            },
        });
    }

    async listNotes(batchId: string) {
        return prisma.batchNote.findMany({
            where: { batchId },
            include: {
                uploadedBy: { select: { id: true, name: true, picture: true } },
                attachments: { orderBy: { createdAt: 'asc' } },
            },
            orderBy: { createdAt: 'desc' },
        });
    }

    async getRecentNotes(userId: string, coachingId: string) {
        // Get user's wards in parallel with nothing (prep step)
        const userWards = await prisma.ward.findMany({
            where: { parentId: userId },
            select: { id: true },
        });
        const wardIds = userWards.map((w) => w.id);

        // Get coaching members for both user + wards
        const whereConditions: any[] = [{ userId }];
        if (wardIds.length > 0) {
            whereConditions.push({ wardId: { in: wardIds } });
        }

        // Single query: get batch IDs through coaching members → batch members
        const memberIds = (await prisma.coachingMember.findMany({
            where: { coachingId, OR: whereConditions },
            select: { id: true },
        })).map((m) => m.id);

        if (memberIds.length === 0) return [];

        const batchIds = (await prisma.batchMember.findMany({
            where: { memberId: { in: memberIds } },
            select: { batchId: true },
        })).map((b) => b.batchId);

        if (batchIds.length === 0) return [];

        // Get notes from last 7 days
        const sevenDaysAgo = new Date();
        sevenDaysAgo.setDate(sevenDaysAgo.getDate() - 7);

        return prisma.batchNote.findMany({
            where: {
                batchId: { in: batchIds },
                createdAt: { gte: sevenDaysAgo },
            },
            include: {
                uploadedBy: { select: { id: true, name: true, picture: true } },
                attachments: { orderBy: { createdAt: 'asc' } },
                batch: { select: { id: true, name: true, subject: true } },
            },
            orderBy: { createdAt: 'desc' },
            take: 20,  // Limit results for performance
        });
    }

    async deleteNote(noteId: string) {
        // Attachments cascade-delete via Prisma relation
        return prisma.batchNote.delete({ where: { id: noteId } });
    }

    // ── Storage tracking ──────────────────────────────────────────────

    /** Add bytes to coaching storage counter */
    async addStorageUsage(coachingId: string, bytes: number) {
        return prisma.coaching.update({
            where: { id: coachingId },
            data: { storageUsed: { increment: bytes } },
            select: { storageUsed: true, storageLimit: true },
        });
    }

    /** Subtract bytes from coaching storage counter */
    async subtractStorageUsage(coachingId: string, bytes: number) {
        return prisma.coaching.update({
            where: { id: coachingId },
            data: { storageUsed: { decrement: bytes } },
            select: { storageUsed: true, storageLimit: true },
        });
    }

    /** Get current storage usage for a coaching */
    async getStorageUsage(coachingId: string) {
        const coaching = await prisma.coaching.findUnique({
            where: { id: coachingId },
            select: { storageUsed: true, storageLimit: true },
        });
        return {
            used: Number(coaching?.storageUsed ?? 0),
            limit: Number(coaching?.storageLimit ?? 524288000),
        };
    }

    /** Get total bytes of attachments for a note (for deletion) */
    async getNoteAttachmentsSize(noteId: string): Promise<number> {
        const attachments = await prisma.noteAttachment.findMany({
            where: { noteId },
            select: { fileSize: true },
        });
        return attachments.reduce((sum, a) => sum + a.fileSize, 0);
    }

    // ── Notices (Announcements) ───────────────────────────────────────

    async createNotice(batchId: string, sentById: string, data: {
        title: string;
        message: string;
        priority?: string;
        type?: string;
        date?: string;
        startTime?: string;
        endTime?: string;
        day?: string;
        location?: string;
    }) {
        const createData: Record<string, any> = {
            batchId,
            sentById,
            title: data.title,
            message: data.message,
        };
        if (data.priority) createData.priority = data.priority;
        if (data.type) createData.type = data.type;
        if (data.date) createData.date = new Date(data.date);
        if (data.startTime) createData.startTime = data.startTime;
        if (data.endTime) createData.endTime = data.endTime;
        if (data.day) createData.day = data.day;
        if (data.location) createData.location = data.location;

        return prisma.batchNotice.create({
            data: createData as any,
            include: {
                sentBy: { select: { id: true, name: true, picture: true } },
            },
        }).then((result) => {
            // Fire notification for new notice
            onNoticeCreated(result.id, data.title, batchId);
            return result;
        });
    }

    async listNotices(batchId: string) {
        return prisma.batchNotice.findMany({
            where: { batchId },
            include: {
                sentBy: { select: { id: true, name: true, picture: true } },
            },
            orderBy: { createdAt: 'desc' },
        });
    }

    async deleteNotice(noticeId: string) {
        return prisma.batchNotice.delete({ where: { id: noticeId } });
    }

    // ── Helpers ───────────────────────────────────────────────────────

    /** Verify a user is an ADMIN/OWNER of the coaching that owns this batch */
    async verifyBatchAdmin(batchId: string, userId: string) {
        const batch = await prisma.batch.findUnique({
            where: { id: batchId },
            select: {
                id: true,
                coachingId: true,
                coaching: { select: { ownerId: true } },
            },
        });
        if (!batch) return null;

        if (batch.coaching.ownerId === userId) return batch;

        const adminMember = await prisma.coachingMember.findFirst({
            where: { coachingId: batch.coachingId, userId, role: 'ADMIN', status: 'active' },
        });
        return adminMember ? batch : null;
    }

    /** Verify a user is a TEACHER assigned to this batch (single query) */
    async verifyBatchTeacher(batchId: string, userId: string) {
        const batch = await prisma.batch.findUnique({
            where: { id: batchId },
            select: {
                id: true,
                coachingId: true,
                coaching: { select: { ownerId: true } },
            },
        });
        if (!batch) return null;

        // Owner has full access
        if (batch.coaching.ownerId === userId) return batch;

        // Single query: check if user is admin OR teacher in batch
        const member = await prisma.coachingMember.findFirst({
            where: {
                coachingId: batch.coachingId,
                userId,
                status: 'active',
                OR: [
                    { role: 'ADMIN' },
                    { batchMembers: { some: { batchId, role: 'TEACHER' } } },
                ],
            },
        });
        return member ? batch : null;
    }

    /** Verify a user has ANY access to this batch (admin, teacher, or student) */
    async verifyBatchAccess(batchId: string, userId: string) {
        const batch = await prisma.batch.findUnique({
            where: { id: batchId },
            select: {
                id: true,
                coachingId: true,
                coaching: { select: { ownerId: true } },
            },
        });
        if (!batch) return null;

        if (batch.coaching.ownerId === userId) return batch;

        // Check if any coaching member
        const member = await prisma.coachingMember.findFirst({
            where: { coachingId: batch.coachingId, userId, status: 'active' },
        });
        return member ? batch : null;
    }

    /** Get batches a specific user belongs to (via their coaching member) */
    async getMyBatches(coachingId: string, userId: string) {
        // First find the user's coaching member ID
        const coachingMember = await prisma.coachingMember.findFirst({
            where: { coachingId, userId, status: 'active' },
        });

        if (!coachingMember) return [];

        return prisma.batch.findMany({
            where: {
                coachingId,
                status: 'active',
                members: { some: { memberId: coachingMember.id } },
            },
            select: BATCH_LIST_SELECT,
            orderBy: { createdAt: 'desc' },
        });
    }

    /**
     * Dashboard feed — recent assessments, assignments, and notices
     * across all batches the user belongs to in a coaching.
     */
    async getDashboardFeed(userId: string, coachingId: string) {
        // Get user's wards
        const userWards = await prisma.ward.findMany({
            where: { parentId: userId },
            select: { id: true },
        });
        const wardIds = userWards.map((w) => w.id);

        const whereConditions: any[] = [{ userId }];
        if (wardIds.length > 0) {
            whereConditions.push({ wardId: { in: wardIds } });
        }

        const memberIds = (await prisma.coachingMember.findMany({
            where: { coachingId, OR: whereConditions },
            select: { id: true },
        })).map((m) => m.id);

        if (memberIds.length === 0) return { assessments: [], assignments: [], notices: [] };

        const batchIds = (await prisma.batchMember.findMany({
            where: { memberId: { in: memberIds } },
            select: { batchId: true },
        })).map((b) => b.batchId);

        if (batchIds.length === 0) return { assessments: [], assignments: [], notices: [] };

        const sevenDaysAgo = new Date();
        sevenDaysAgo.setDate(sevenDaysAgo.getDate() - 7);

        // Fetch all three in parallel
        const [assessments, assignments, notices] = await Promise.all([
            // Recent published assessments
            prisma.assessment.findMany({
                where: {
                    batchId: { in: batchIds },
                    status: 'PUBLISHED',
                    createdAt: { gte: sevenDaysAgo },
                },
                select: {
                    id: true,
                    title: true,
                    type: true,
                    totalMarks: true,
                    durationMinutes: true,
                    createdAt: true,
                    batchId: true,
                    batch: { select: { id: true, name: true } },
                    _count: { select: { questions: true } },
                },
                orderBy: { createdAt: 'desc' },
                take: 10,
            }),
            // Recent assignments
            prisma.assignment.findMany({
                where: {
                    batchId: { in: batchIds },
                    status: 'ACTIVE',
                    createdAt: { gte: sevenDaysAgo },
                },
                select: {
                    id: true,
                    title: true,
                    dueDate: true,
                    totalMarks: true,
                    createdAt: true,
                    batchId: true,
                    batch: { select: { id: true, name: true } },
                },
                orderBy: { createdAt: 'desc' },
                take: 10,
            }),
            // Recent notices
            prisma.batchNotice.findMany({
                where: {
                    batchId: { in: batchIds },
                    createdAt: { gte: sevenDaysAgo },
                },
                select: {
                    id: true,
                    title: true,
                    message: true,
                    createdAt: true,
                    batchId: true,
                    batch: { select: { id: true, name: true } },
                    sentBy: { select: { id: true, name: true } },
                },
                orderBy: { createdAt: 'desc' },
                take: 10,
            }),
        ]);

        return { assessments, assignments, notices };
    }
}

File: frontend/lib/features/batch/services/batch_service.dart
import '../../../core/constants/api_constants.dart';
import '../../../core/services/cache_manager.dart';
import '../../../shared/services/api_client.dart';
import '../models/batch_model.dart';
import '../models/batch_member_model.dart';
import '../models/batch_note_model.dart';
import '../models/batch_notice_model.dart';

class BatchService {
  final ApiClient _api = ApiClient.instance;
  final CacheManager _cache = CacheManager.instance;

  // ── File Upload ───────────────────────────────────────────────────

  /// Upload a single note file, returns the URL and metadata.
  Future<Map<String, dynamic>> uploadNoteFile(String filePath) async {
    return await _api.uploadFile(
      ApiConstants.uploadNote,
      fieldName: 'file',
      filePath: filePath,
    );
  }

  /// Upload multiple note files in one request.
  /// Returns { files: [...], totalSize: int }.
  Future<Map<String, dynamic>> uploadNoteFiles(List<String> filePaths) async {
    return await _api.uploadFiles(
      ApiConstants.uploadNotes,
      fieldName: 'files',
      filePaths: filePaths,
    );
  }

  /// GET /coaching/:coachingId/batches/storage
  Future<Map<String, dynamic>> getStorageUsage(String coachingId) async {
    return await _api.getAuthenticated(ApiConstants.batchStorage(coachingId));
  }

  // ── CRUD ──────────────────────────────────────────────────────────

  /// POST /coaching/:coachingId/batches
  Future<BatchModel> createBatch(
    String coachingId, {
    required String name,
    String? subject,
    String? description,
    String? startTime,
    String? endTime,
    List<String>? days,
    int? maxStudents,
  }) async {
    final body = <String, dynamic>{'name': name};
    if (subject != null) body['subject'] = subject;
    if (description != null) body['description'] = description;
    if (startTime != null) body['startTime'] = startTime;
    if (endTime != null) body['endTime'] = endTime;
    if (days != null) body['days'] = days;
    if (maxStudents != null) body['maxStudents'] = maxStudents;

    final data = await _api.postAuthenticated(
      ApiConstants.batches(coachingId),
      body: body,
    );
    await _cache.invalidatePrefix('batch:$coachingId');
    return BatchModel.fromJson(data['batch'] as Map<String, dynamic>);
  }

  /// GET /coaching/:coachingId/batches
  Future<List<BatchModel>> listBatches(
    String coachingId, {
    String? status,
  }) async {
    final key = 'batch:$coachingId:list${status != null ? ':$status' : ''}';
    try {
      var url = ApiConstants.batches(coachingId);
      if (status != null) url += '?status=$status';
      final data = await _api.getAuthenticated(url);
      await _cache.put(key, data);
      final list = data['batches'] as List<dynamic>;
      return list
          .map((e) => BatchModel.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      final cached = await _cache.get(key);
      if (cached != null) {
        final list = (cached['batches'] as List<dynamic>);
        return list
            .map((e) =>
                BatchModel.fromJson(Map<String, dynamic>.from(e as Map)))
            .toList();
      }
      rethrow;
    }
  }

  /// GET /coaching/:coachingId/batches/my
  Future<List<BatchModel>> getMyBatches(String coachingId) async {
    final key = 'batch:$coachingId:my';
    try {
      final data = await _api.getAuthenticated(
        ApiConstants.myBatches(coachingId),
      );
      await _cache.put(key, data);
      final list = data['batches'] as List<dynamic>;
      return list
          .map((e) => BatchModel.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      final cached = await _cache.get(key);
      if (cached != null) {
        final list = (cached['batches'] as List<dynamic>);
        return list
            .map((e) =>
                BatchModel.fromJson(Map<String, dynamic>.from(e as Map)))
            .toList();
      }
      rethrow;
    }
  }

  /// GET /coaching/:coachingId/batches/:batchId
  Future<BatchModel> getBatchById(String coachingId, String batchId) async {
    final key = 'batch:$coachingId:$batchId';
    try {
      final data = await _api.getAuthenticated(
        ApiConstants.batchById(coachingId, batchId),
      );
      await _cache.put(key, data);
      return BatchModel.fromJson(data['batch'] as Map<String, dynamic>);
    } catch (e) {
      final cached = await _cache.get(key);
      if (cached != null) {
        return BatchModel.fromJson(
            Map<String, dynamic>.from(cached['batch'] as Map));
      }
      rethrow;
    }
  }

  /// PATCH /coaching/:coachingId/batches/:batchId
  Future<BatchModel> updateBatch(
    String coachingId,
    String batchId, {
    String? name,
    String? subject,
    String? description,
    String? startTime,
    String? endTime,
    List<String>? days,
    int? maxStudents,
    String? status,
  }) async {
    final body = <String, dynamic>{};
    if (name != null) body['name'] = name;
    if (subject != null) body['subject'] = subject;
    if (description != null) body['description'] = description;
    if (startTime != null) body['startTime'] = startTime;
    if (endTime != null) body['endTime'] = endTime;
    if (days != null) body['days'] = days;
    if (maxStudents != null) body['maxStudents'] = maxStudents;
    if (status != null) body['status'] = status;

    final data = await _api.patchAuthenticated(
      ApiConstants.batchById(coachingId, batchId),
      body: body,
    );
    await _cache.invalidatePrefix('batch:$coachingId');
    return BatchModel.fromJson(data['batch'] as Map<String, dynamic>);
  }

  /// DELETE /coaching/:coachingId/batches/:batchId
  Future<bool> deleteBatch(String coachingId, String batchId) async {
    final ok =
        await _api.deleteAuthenticated(ApiConstants.batchById(coachingId, batchId));
    if (ok) await _cache.invalidatePrefix('batch:$coachingId');
    return ok;
  }

  // ── Members ───────────────────────────────────────────────────────

  /// POST /coaching/:coachingId/batches/:batchId/members
  Future<List<BatchMemberModel>> addMembers(
    String coachingId,
    String batchId, {
    required List<String> memberIds,
    String role = 'STUDENT',
  }) async {
    final data = await _api.postAuthenticated(
      ApiConstants.batchMembers(coachingId, batchId),
      body: {'memberIds': memberIds, 'role': role},
    );
    await _cache.invalidate('batch:$coachingId:$batchId:members');
    final list = data['members'] as List<dynamic>;
    return list
        .map((e) => BatchMemberModel.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// GET /coaching/:coachingId/batches/:batchId/members
  Future<List<BatchMemberModel>> getMembers(
    String coachingId,
    String batchId,
  ) async {
    final key = 'batch:$coachingId:$batchId:members';
    try {
      final data = await _api.getAuthenticated(
        ApiConstants.batchMembers(coachingId, batchId),
      );
      await _cache.put(key, data);
      final list = data['members'] as List<dynamic>;
      return list
          .map((e) => BatchMemberModel.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      final cached = await _cache.get(key);
      if (cached != null) {
        final list = (cached['members'] as List<dynamic>);
        return list
            .map((e) => BatchMemberModel.fromJson(
                Map<String, dynamic>.from(e as Map)))
            .toList();
      }
      rethrow;
    }
  }

  /// GET /coaching/:coachingId/batches/:batchId/members/available?role=...
  Future<List<dynamic>> getAvailableMembers(
    String coachingId,
    String batchId, {
    String? role,
  }) async {
    var url = ApiConstants.batchAvailableMembers(coachingId, batchId);
    if (role != null) url += '?role=$role';
    final data = await _api.getAuthenticated(url);
    return data['members'] as List<dynamic>;
  }

  /// DELETE /coaching/:coachingId/batches/:batchId/members/:batchMemberId
  Future<bool> removeMember(
    String coachingId,
    String batchId,
    String batchMemberId,
  ) async {
    final ok = await _api.deleteAuthenticated(
      ApiConstants.removeBatchMember(coachingId, batchId, batchMemberId),
    );
    if (ok) await _cache.invalidate('batch:$coachingId:$batchId:members');
    return ok;
  }

  // ── Notes ─────────────────────────────────────────────────────────

  /// POST /coaching/:coachingId/batches/:batchId/notes
  Future<BatchNoteModel> createNote(
    String coachingId,
    String batchId, {
    required String title,
    String? description,
    List<Map<String, dynamic>>? attachments,
  }) async {
    final body = <String, dynamic>{'title': title};
    if (description != null) body['description'] = description;
    if (attachments != null && attachments.isNotEmpty) {
      body['attachments'] = attachments;
    }

    final data = await _api.postAuthenticated(
      ApiConstants.batchNotes(coachingId, batchId),
      body: body,
    );
    await Future.wait([
      _cache.invalidate('batch:$coachingId:$batchId:notes'),
      _cache.invalidate('batch:$coachingId:recent-notes'),
    ]);
    return BatchNoteModel.fromJson(data['note'] as Map<String, dynamic>);
  }

  /// GET /coaching/:coachingId/batches/:batchId/notes
  Future<List<BatchNoteModel>> listNotes(
    String coachingId,
    String batchId,
  ) async {
    final key = 'batch:$coachingId:$batchId:notes';
    try {
      final data = await _api.getAuthenticated(
        ApiConstants.batchNotes(coachingId, batchId),
      );
      await _cache.put(key, data);
      final list = data['notes'] as List<dynamic>;
      return list
          .map((e) => BatchNoteModel.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      final cached = await _cache.get(key);
      if (cached != null) {
        final list = (cached['notes'] as List<dynamic>);
        return list
            .map((e) => BatchNoteModel.fromJson(
                Map<String, dynamic>.from(e as Map)))
            .toList();
      }
      rethrow;
    }
  }

  /// GET /coaching/:coachingId/batches/recent-notes
  Future<List<BatchNoteModel>> getRecentNotes(String coachingId) async {
    final key = 'batch:$coachingId:recent-notes';
    try {
      final data = await _api.getAuthenticated(
        ApiConstants.recentNotes(coachingId),
      );
      await _cache.put(key, data);
      final list = data['notes'] as List<dynamic>;
      return list
          .map((e) => BatchNoteModel.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      final cached = await _cache.get(key);
      if (cached != null) {
        final list = (cached['notes'] as List<dynamic>);
        return list
            .map((e) => BatchNoteModel.fromJson(
                Map<String, dynamic>.from(e as Map)))
            .toList();
      }
      return [];
    }
  }

  /// DELETE /coaching/:coachingId/batches/:batchId/notes/:noteId
  Future<bool> deleteNote(String coachingId, String batchId, String noteId) async {
    final ok = await _api.deleteAuthenticated(
      ApiConstants.deleteBatchNote(coachingId, batchId, noteId),
    );
    if (ok) {
      await Future.wait([
        _cache.invalidate('batch:$coachingId:$batchId:notes'),
        _cache.invalidate('batch:$coachingId:recent-notes'),
      ]);
    }
    return ok;
  }

  // ── Notices ───────────────────────────────────────────────────────

  /// POST /coaching/:coachingId/batches/:batchId/notices
  Future<BatchNoticeModel> createNotice(
    String coachingId,
    String batchId, {
    required String title,
    required String message,
    String priority = 'normal',
    String type = 'general',
    String? date,
    String? startTime,
    String? endTime,
    String? day,
    String? location,
  }) async {
    final body = <String, dynamic>{
      'title': title,
      'message': message,
      'priority': priority,
      'type': type,
    };
    if (date != null) body['date'] = date;
    if (startTime != null) body['startTime'] = startTime;
    if (endTime != null) body['endTime'] = endTime;
    if (day != null) body['day'] = day;
    if (location != null) body['location'] = location;

    final data = await _api.postAuthenticated(
      ApiConstants.batchNotices(coachingId, batchId),
      body: body,
    );
    await _cache.invalidate('batch:$coachingId:$batchId:notices');
    return BatchNoticeModel.fromJson(data['notice'] as Map<String, dynamic>);
  }

  /// GET /coaching/:coachingId/batches/:batchId/notices
  Future<List<BatchNoticeModel>> listNotices(
    String coachingId,
    String batchId,
  ) async {
    final key = 'batch:$coachingId:$batchId:notices';
    try {
      final data = await _api.getAuthenticated(
        ApiConstants.batchNotices(coachingId, batchId),
      );
      await _cache.put(key, data);
      final list = data['notices'] as List<dynamic>;
      return list
          .map((e) => BatchNoticeModel.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      final cached = await _cache.get(key);
      if (cached != null) {
        final list = (cached['notices'] as List<dynamic>);
        return list
            .map((e) => BatchNoticeModel.fromJson(
                Map<String, dynamic>.from(e as Map)))
            .toList();
      }
      rethrow;
    }
  }

  /// DELETE /coaching/:coachingId/batches/:batchId/notices/:noticeId
  Future<bool> deleteNotice(
    String coachingId,
    String batchId,
    String noticeId,
  ) async {
    final ok = await _api.deleteAuthenticated(
      ApiConstants.deleteBatchNotice(coachingId, batchId, noticeId),
    );
    if (ok) await _cache.invalidate('batch:$coachingId:$batchId:notices');
    return ok;
  }
}
